diff --git a/CMakeLists.txt b/CMakeLists.txt
index 9eca7a696..048570a9c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -493,7 +493,7 @@ if(FLB_TLS)
   add_subdirectory(${FLB_PATH_LIB_MBEDTLS} EXCLUDE_FROM_ALL)
 
   # Link OpenSSL statically on Windows.
-  if (FLB_SYSTEM_WINDOWS)
+  if (FLB_SYSTEM_WINDOWS AND MSVC)
     set(OPENSSL_USE_STATIC_LIBS ON)
     set(OPENSSL_MSVC_STATIC_RT  ON)
   endif()
diff --git a/cmake/luajit.cmake b/cmake/luajit.cmake
index 7741805a9..63d697510 100644
--- a/cmake/luajit.cmake
+++ b/cmake/luajit.cmake
@@ -11,6 +11,10 @@ add_library(libluajit STATIC IMPORTED GLOBAL)
 set(LUAJIT_SRC ${CMAKE_CURRENT_SOURCE_DIR}/${FLB_PATH_LIB_LUAJIT})
 set(LUAJIT_DEST ${CMAKE_CURRENT_BINARY_DIR})
 
+if(NOT DEFINED CMAKE_HOST_C_COMPILER)
+  set(CMAKE_HOST_C_COMPILER, ${CMAKE_C_COMPILER})
+endif()
+
 if (CMAKE_SYSTEM_NAME MATCHES "Darwin")
   set(CFLAGS "${CFLAGS} -isysroot ${CMAKE_OSX_SYSROOT} -fno-stack-check")
 
@@ -30,6 +34,10 @@ else()
   set(DEPLOYMENT_TARGET "")
 endif()
 
+if (CMAKE_SYSTEM_NAME MATCHES "Windows")
+  set(TARGET_SYS_PARAM "TARGET_SYS=${CMAKE_SYSTEM_NAME}")
+endif()
+
 # luajit (UNIX)
 # =============
 ExternalProject_Add(luajit
@@ -37,8 +45,8 @@ ExternalProject_Add(luajit
   EXCLUDE_FROM_ALL TRUE
   SOURCE_DIR ${LUAJIT_SRC}
   CONFIGURE_COMMAND ./configure
-  BUILD_COMMAND $(MAKE) CC=${CMAKE_C_COMPILER} ${DEPLOYMENT_TARGET} CFLAGS=${CFLAGS} BUILDMODE=static "XCFLAGS=-fPIC"
-  INSTALL_COMMAND cp src/libluajit.a "${LUAJIT_DEST}/lib/libluajit.a")
+  BUILD_COMMAND $(MAKE) ${TARGET_SYS_PARAM} HOST_CC=${CMAKE_HOST_C_COMPILER} CC=${CMAKE_C_COMPILER} ${DEPLOYMENT_TARGET} CFLAGS=${CFLAGS} BUILDMODE=static "XCFLAGS=-fPIC" Q=
+  INSTALL_COMMAND ${CMAKE_COMMAND} -E copy src/libluajit.a "${LUAJIT_DEST}/lib/libluajit.a")
 
 # luajit (Windows)
 # ================
diff --git a/include/fluent-bit/flb_compat.h b/include/fluent-bit/flb_compat.h
index 37f3f9c66..6d3638ac4 100644
--- a/include/fluent-bit/flb_compat.h
+++ b/include/fluent-bit/flb_compat.h
@@ -27,17 +27,12 @@
 #ifndef FLB_COMPAT_H
 #define FLB_COMPAT_H
 
-/*
- * libmonkey exposes compat macros for <unistd.h>, which some platforms lack,
- * so include the header here.
- */
-#include <monkey/mk_core.h>
-
 #ifdef FLB_SYSTEM_WINDOWS
 #define WIN32_LEAN_AND_MEAN
 #include <winsock2.h>
+#include <ws2tcpip.h>
 #include <windows.h>
-#include <Wincrypt.h> /* flb_io_tls.c */
+#include <wincrypt.h> /* flb_io_tls.c */
 
 #include <monkey/mk_core/mk_sleep.h>
 #include <fluent-bit/flb_dlfcn_win32.h>
@@ -122,11 +117,13 @@ static inline char* realpath(char *path, char *buf)
     return _fullpath(NULL, path, 0);
 }
 
+#ifdef _MSC_VER
 static inline int usleep(LONGLONG usec)
 {
     // Convert into 100ns unit.
     return nanosleep(usec * 10);
 }
+#endif
 #else
 #include <netdb.h>
 #include <netinet/in.h>
@@ -139,4 +136,10 @@ static inline int usleep(LONGLONG usec)
 #define FLB_DIRCHAR '/'
 #endif
 
+/*
+ * libmonkey exposes compat macros for <unistd.h>, which some platforms lack,
+ * so include the header here.
+ */
+#include <monkey/mk_core.h>
+
 #endif
diff --git a/include/fluent-bit/flb_langinfo.h b/include/fluent-bit/flb_langinfo.h
index 6da306829..54331bc9b 100644
--- a/include/fluent-bit/flb_langinfo.h
+++ b/include/fluent-bit/flb_langinfo.h
@@ -26,7 +26,7 @@
 #ifndef FLB_LANGINFO_H
 #define FLB_LANGINFO_H
 
-#ifndef _MSC_VER
+#ifndef _WIN32
 #include <langinfo.h>
 #else
 
diff --git a/lib/avro/src/errors.c b/lib/avro/src/errors.c
index 8abd8c832..0eabedf23 100644
--- a/lib/avro/src/errors.c
+++ b/lib/avro/src/errors.c
@@ -52,7 +52,7 @@ static void make_error_data_key()
     pthread_key_create(&error_data_key, free);
 }
 #elif defined _WIN32
-#include <Windows.h>
+#include <windows.h>
 
 static __declspec( thread ) struct avro_error_data_t TLS_ERROR_DATA = { "", "", NULL, NULL };
 
diff --git a/lib/c-ares-809d5e84/CMakeLists.txt b/lib/c-ares-809d5e84/CMakeLists.txt
index a423b5a2a..5ae46dfcc 100644
--- a/lib/c-ares-809d5e84/CMakeLists.txt
+++ b/lib/c-ares-809d5e84/CMakeLists.txt
@@ -233,7 +233,7 @@ IF (HAVE_LIBRT)
 	LIST (APPEND CARES_DEPENDENT_LIBS rt)
 ENDIF ()
 IF (WIN32)
-	LIST (APPEND CARES_DEPENDENT_LIBS ws2_32 Advapi32)
+	LIST (APPEND CARES_DEPENDENT_LIBS ws2_32 advapi32)
 ENDIF ()
 
 
diff --git a/lib/chunkio/include/chunkio/chunkio_compat.h b/lib/chunkio/include/chunkio/chunkio_compat.h
index 3e4951b98..0d4c5d2d6 100644
--- a/lib/chunkio/include/chunkio/chunkio_compat.h
+++ b/lib/chunkio/include/chunkio/chunkio_compat.h
@@ -32,6 +32,7 @@
 #include <direct.h>
 #pragma comment(lib, "ws2_32.lib")
 
+#ifndef __MINGW32__
 /** mode flags for access() */
 #define R_OK 04
 #define W_OK 02
@@ -40,10 +41,12 @@
 
 #define PATH_MAX MAX_PATH
 #define S_ISREG(m) (((m) & S_IFMT) == S_IFREG)
-#define strerror_r(errno,buf,len) strerror_s(buf,len,errno)
 
 typedef SSIZE_T ssize_t;
 typedef unsigned mode_t;
+#endif
+
+#define strerror_r(errno,buf,len) strerror_s(buf,len,errno)
 
 static inline char* dirname(const char *path)
 {
diff --git a/lib/chunkio/src/CMakeLists.txt b/lib/chunkio/src/CMakeLists.txt
index a4fc2d37b..db58e1add 100644
--- a/lib/chunkio/src/CMakeLists.txt
+++ b/lib/chunkio/src/CMakeLists.txt
@@ -22,8 +22,8 @@ if(${CMAKE_SYSTEM_NAME} MATCHES "Windows")
     )
   set(libs
     ${libs}
-    Shell32.lib
-    Shlwapi.lib)
+    shell32.lib
+    shlwapi.lib)
 else()
   set(src
     ${src}
diff --git a/lib/chunkio/src/cio_os.c b/lib/chunkio/src/cio_os.c
index b368215fa..140aeea77 100644
--- a/lib/chunkio/src/cio_os.c
+++ b/lib/chunkio/src/cio_os.c
@@ -27,7 +27,7 @@
 
 #include <chunkio/chunkio_compat.h>
 
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(__MINGW32__)
 #include <Shlobj.h>
 #endif
 
diff --git a/lib/chunkio/src/cio_utils.c b/lib/chunkio/src/cio_utils.c
index 16d4af3da..e297d8d1d 100644
--- a/lib/chunkio/src/cio_utils.c
+++ b/lib/chunkio/src/cio_utils.c
@@ -25,7 +25,7 @@
 #include <sys/stat.h>
 #include <fcntl.h>
 #include <errno.h>
-#ifndef _MSC_VER
+#ifndef _WIN32
 #include <fts.h>
 #endif
 
@@ -33,7 +33,7 @@
 #include <chunkio/chunkio_compat.h>
 #include <chunkio/cio_log.h>
 
-#ifndef _MSC_VER
+#ifndef _WIN32
 /*
  * Taken from StackOverflow:
  *
diff --git a/lib/chunkio/src/win32/dirent.c b/lib/chunkio/src/win32/dirent.c
index 2ed39ed15..7d1a6584f 100644
--- a/lib/chunkio/src/win32/dirent.c
+++ b/lib/chunkio/src/win32/dirent.c
@@ -22,7 +22,7 @@
  * Win32's FIndFirstFile/FindNextFile API.
  */
 
-#include <Windows.h>
+#include <windows.h>
 #include <shlwapi.h>
 
 #include "dirent.h"
diff --git a/lib/monkey/include/monkey/mk_core/external/winpthreads.h b/lib/monkey/include/monkey/mk_core/external/winpthreads.h
index 97045a69a..f28366d2c 100644
--- a/lib/monkey/include/monkey/mk_core/external/winpthreads.h
+++ b/lib/monkey/include/monkey/mk_core/external/winpthreads.h
@@ -101,7 +101,7 @@ extern "C" {
 #define PTHREAD_BARRIER_SERIAL_THREAD 1
 
 /* Windows doesn't have this, so declare it ourselves. */
-#if (_MSC_VER < 1900)
+#if defined(_MSC_VER) && (_MSC_VER < 1900)
 struct timespec
 {
   /* long long in windows is the same as long in unix for 64bit */
diff --git a/lib/monkey/include/monkey/mk_core/external/winuio.h b/lib/monkey/include/monkey/mk_core/external/winuio.h
index 4ae747888..5c61df6ac 100644
--- a/lib/monkey/include/monkey/mk_core/external/winuio.h
+++ b/lib/monkey/include/monkey/mk_core/external/winuio.h
@@ -8,7 +8,7 @@
 #else
 #include <errno.h>
 #include <io.h>
-#include <BaseTsd.h>
+#include <basetsd.h>
 #include <winsock2.h>
 #include <ws2tcpip.h>
 typedef SSIZE_T ssize_t;
diff --git a/lib/monkey/include/monkey/mk_core/mk_string.h b/lib/monkey/include/monkey/mk_core/mk_string.h
index cdbe6d505..65c032de9 100644
--- a/lib/monkey/include/monkey/mk_core/mk_string.h
+++ b/lib/monkey/include/monkey/mk_core/mk_string.h
@@ -26,7 +26,7 @@
 #include "mk_list.h"
 #include "mk_macros.h"
 
-#if defined(_WIN32) || defined(_WIN64)
+#if (defined(_WIN32) || defined(_WIN64)) && !defined(__MINGW32__)
   #define snprintf _snprintf
   #define vsnprintf _vsnprintf
   #define strcasecmp _stricmp
diff --git a/lib/monkey/mk_core/deps/libevent/cmake/CheckFunctionExistsEx.c b/lib/monkey/mk_core/deps/libevent/cmake/CheckFunctionExistsEx.c
index 5ee3e5913..84bae9feb 100644
--- a/lib/monkey/mk_core/deps/libevent/cmake/CheckFunctionExistsEx.c
+++ b/lib/monkey/mk_core/deps/libevent/cmake/CheckFunctionExistsEx.c
@@ -11,7 +11,7 @@ int main(){
 #else
 int main(int ac, char*av[]){
 #endif
-#ifdef _WIN32
+#if defined(_WIN32) && !defined(__MINGW32__)
   void * p = &CHECK_FUNCTION_EXISTS;
 #else
   CHECK_FUNCTION_EXISTS();
diff --git a/lib/monkey/mk_core/mk_rconf.c b/lib/monkey/mk_core/mk_rconf.c
index 621ff2d3a..f57c60a61 100644
--- a/lib/monkey/mk_core/mk_rconf.c
+++ b/lib/monkey/mk_core/mk_rconf.c
@@ -23,7 +23,7 @@
 #include <string.h>
 #include <sys/types.h>
 #include <sys/stat.h>
-#ifndef _MSC_VER
+#ifndef _WIN32
 #include <glob.h>
 #endif
 
@@ -33,7 +33,7 @@
 #include <mk_core/mk_list.h>
 
 #ifdef _WIN32
-#include <Windows.h>
+#include <windows.h>
 #include <strsafe.h>
 #define PATH_MAX MAX_PATH
 #endif
@@ -589,7 +589,7 @@ static int mk_rconf_path_set(struct mk_rconf *conf, char *file)
     char *end;
     char path[PATH_MAX + 1];
 
-#ifdef _MSC_VER
+#ifdef _WIN32
     p = _fullpath(path, file, PATH_MAX + 1);
 #else
     p = realpath(file, path);
diff --git a/plugins/CMakeLists.txt b/plugins/CMakeLists.txt
index 12f27bcdb..295c6b4f2 100644
--- a/plugins/CMakeLists.txt
+++ b/plugins/CMakeLists.txt
@@ -201,7 +201,9 @@ if (FLB_SYSTEM_WINDOWS)
   REGISTER_IN_PLUGIN("in_winstat")
   REGISTER_IN_PLUGIN("in_winevtlog")
   REGISTER_IN_PLUGIN("in_windows_exporter_metrics")
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W2")
+  if (MSVC)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W2")
+  endif()
 else()
   REGISTER_IN_PLUGIN("in_serial")
 endif()
diff --git a/plugins/in_tail/CMakeLists.txt b/plugins/in_tail/CMakeLists.txt
index 31d865218..83ee195c4 100644
--- a/plugins/in_tail/CMakeLists.txt
+++ b/plugins/in_tail/CMakeLists.txt
@@ -25,13 +25,13 @@ if(FLB_PARSER)
     )
 endif()
 
-if(MSVC)
+if(FLB_SYSTEM_WINDOWS)
   set(src
     ${src}
     win32/stat.c
     win32/io.c
     )
-  FLB_PLUGIN(in_tail "${src}" "Shlwapi")
+  FLB_PLUGIN(in_tail "${src}" "shlwapi")
 else()
   FLB_PLUGIN(in_tail "${src}" "")
 endif()
diff --git a/plugins/in_tail/tail_file.c b/plugins/in_tail/tail_file.c
index 89a8c525b..a144be33f 100644
--- a/plugins/in_tail/tail_file.c
+++ b/plugins/in_tail/tail_file.c
@@ -40,10 +40,6 @@
 #include "tail_multiline.h"
 #include "tail_scan.h"
 
-#ifdef FLB_SYSTEM_WINDOWS
-#include "win32.h"
-#endif
-
 #include <xxhash.h>
 
 static inline void consume_bytes(char *buf, int bytes, int length)
diff --git a/plugins/in_tail/win32/io.c b/plugins/in_tail/win32/io.c
index 45928b04a..9336d1690 100644
--- a/plugins/in_tail/win32/io.c
+++ b/plugins/in_tail/win32/io.c
@@ -17,7 +17,7 @@
  *  limitations under the License.
  */
 
-#include <Windows.h>
+#include <windows.h>
 #include <stdlib.h>
 #include <stdint.h>
 #include <fcntl.h>
diff --git a/plugins/in_tail/win32/stat.c b/plugins/in_tail/win32/stat.c
index c2b14605a..52dce6473 100644
--- a/plugins/in_tail/win32/stat.c
+++ b/plugins/in_tail/win32/stat.c
@@ -17,7 +17,8 @@
  *  limitations under the License.
  */
 
-#include <Windows.h>
+#define _WIN32_WINNT _WIN32_WINNT_WS08
+#include <windows.h>
 #include <stdlib.h>
 #include <stdint.h>
 #include <io.h>
diff --git a/plugins/in_winlog/pack.c b/plugins/in_winlog/pack.c
index 4547d5514..896cb535c 100644
--- a/plugins/in_winlog/pack.c
+++ b/plugins/in_winlog/pack.c
@@ -24,7 +24,6 @@
 #include <fluent-bit/flb_input_plugin.h>
 #include <msgpack.h>
 #include <sddl.h>
-#include <locale.h>
 #include "winlog.h"
 
 #define REGKEY_MAXLEN 256
@@ -84,25 +83,17 @@ static int pack_time(msgpack_packer *mp_pck, int time)
     size_t len;
     struct tm tm;
     char buf[64];
-    _locale_t locale;
 
     if (_localtime32_s(&tm, &time)) {
         flb_errno();
         return -1;
     }
 
-    locale = _get_current_locale();
-    if (locale == NULL) {
-        return -1;
-    }
-
-    len = _strftime_l(buf, 64, FMT_ISO8601, &tm, locale);
+    len = strftime(buf, 64, FMT_ISO8601, &tm);
     if (len == 0) {
         flb_errno();
-        _free_locale(locale);
         return -1;
     }
-    _free_locale(locale);
     msgpack_pack_str(mp_pck, len);
     msgpack_pack_str_body(mp_pck, buf, len);
 
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 630f875bf..209aa6243 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -90,7 +90,9 @@ if(FLB_SYSTEM_WINDOWS)
     ${src}
     flb_dlfcn_win32.c
     )
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W2")
+  if(MSVC)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W2")
+  endif()
 endif()
 
 if(FLB_PARSER)
@@ -236,12 +238,21 @@ check_symbol_exists(accept4 "sys/socket.h" HAVE_ACCEPT4)
 
 # Core dependencies
 if(FLB_SYSTEM_WINDOWS)
-  set(FLB_DEPS
-    "ws2_32.lib"
-    "crypt32.lib"
-    "Bcrypt.lib"
-    "Shlwapi.lib"
-    )
+  if(MSVC)
+    set(FLB_DEPS
+      "ws2_32.lib"
+      "crypt32.lib"
+      "bcrypt.lib"
+      "shlwapi.lib"
+      )
+  else()
+    set(FLB_DEPS
+      "ws2_32.dll"
+      "crypt32.dll"
+      "bcrypt.dll"
+      "shlwapi.dll"
+      )
+  endif()
 else()
   set(FLB_DEPS
     ${FLB_DEPS}
@@ -397,12 +408,17 @@ if(FLB_SHARED_LIB)
     PROPERTIES OUTPUT_NAME fluent-bit)
 
   # Windows doesn't provide pthread (see winpthreads.c in mk_core).
-  if(CMAKE_SYSTEM_NAME MATCHES "Windows")
+  if(MSVC)
     target_link_libraries(fluent-bit-shared ${FLB_DEPS})
   else()
     target_link_libraries(fluent-bit-shared ${FLB_DEPS} -lpthread)
   endif()
 
+  if(FLB_SYSTEM_WINDOWS)
+    # Remove the default "lib" prefix on Windows.
+    set_target_properties(fluent-bit-shared PROPERTIES PREFIX "")
+  endif()
+
   if (MSVC)
     set_target_properties(fluent-bit-shared
       PROPERTIES PDB_NAME fluent-bit.dll)
diff --git a/src/flb_hash.c b/src/flb_hash.c
index fe3035f90..92e764671 100644
--- a/src/flb_hash.c
+++ b/src/flb_hash.c
@@ -24,6 +24,10 @@
 #include <fluent-bit/flb_str.h>
 #include <xxhash.h>
 
+#if defined(_WIN32) && defined(__MINGW32__)
+#define random() rand()
+#endif
+
 static inline void flb_hash_entry_free(struct flb_hash *ht,
                                        struct flb_hash_entry *entry)
 {
diff --git a/src/flb_network.c b/src/flb_network.c
index 992eb1db0..6a89d26bd 100644
--- a/src/flb_network.c
+++ b/src/flb_network.c
@@ -28,11 +28,18 @@
 #include <ctype.h>
 
 #ifdef FLB_SYSTEM_WINDOWS
+#include <sdkddkver.h>
+#define _WIN32_WINNT _WIN32_WINNT_WS08
+#include <winsock2.h>
 #define poll WSAPoll
 #else
 #include <sys/poll.h>
 #endif
 
+#if defined(FLB_SYSTEM_WINDOWS) && !defined(AI_NUMERICSERV)
+#define AI_NUMERICSERV 0x0400
+#endif
+
 #include <fluent-bit/flb_info.h>
 #include <fluent-bit/flb_compat.h>
 #include <fluent-bit/flb_info.h>
diff --git a/src/flb_sosreport.c b/src/flb_sosreport.c
index 7bbd79aef..2768163f1 100644
--- a/src/flb_sosreport.c
+++ b/src/flb_sosreport.c
@@ -28,7 +28,7 @@
 #include <fluent-bit/flb_utils.h>
 #include <fluent-bit/flb_kv.h>
 
-#ifndef _MSC_VER
+#ifndef _WIN32
 #include <sys/utsname.h>
 #endif
 
@@ -107,7 +107,7 @@ static void print_properties(struct mk_list *props)
     }
 }
 
-#ifdef _MSC_VER
+#ifdef _WIN32
 /* A definition table of SYSTEM_INFO.wProcessorArchitecture.
  *
  * This is a streight-forward translation of the official manual.
@@ -188,7 +188,7 @@ int flb_sosreport(struct flb_config *config)
     printf("    Built Flags\t\t%s\n", FLB_INFO_FLAGS);
     printf("\n");
 
-#ifndef _MSC_VER
+#ifndef _WIN32
     struct utsname uts;
     uname(&uts);
 
diff --git a/src/win32/winsvc.c b/src/win32/winsvc.c
index 9e0942ee5..b82f2780b 100644
--- a/src/win32/winsvc.c
+++ b/src/win32/winsvc.c
@@ -17,8 +17,8 @@
  *  limitations under the License.
  */
 
-#include <Windows.h>
-#include <Shlwapi.h>
+#include <windows.h>
+#include <shlwapi.h>
 
 struct flb_config;
 extern struct flb_config *config;
